src/controller/java/AndroidDeviceControllerWrapper.cpp

```diff
@@ -226,3 +280,98 @@ void AndroidDeviceControllerWrapper::OnPairingDeleted(CHIP_ERROR error)
 {
     CallVoidInt(GetJavaEnv(), mJavaObjectRef, "onPairingDeleted", static_cast<jint>(error));
 }
+
+void AndroidDeviceControllerWrapper::OnMessage(chip::System::PacketBufferHandle msg) {}
+
+void AndroidDeviceControllerWrapper::OnStatusChange(void) {}
+
+void AndroidDeviceControllerWrapper::SetDelegate(chip::PersistentStorageResultDelegate * delegate)
+{
+    mStorageResultDelegate = delegate;
+}
+
+void AndroidDeviceControllerWrapper::GetKeyValue(const char * key)
+{
+    jstring keyString = NULL;
+    CHIP_ERROR err    = CHIP_NO_ERROR;
+    jclass storageCls = GetPersistentStorageClass();
+    jmethodID method  = GetJavaEnv()->GetStaticMethodID(storageCls, "getKeyValue", "(Ljava/lang/String;)Ljava/lang/String;");
+
+    GetJavaEnv()->ExceptionClear();
+
+    err = N2J_NewStringUTF(GetJavaEnv(), key, keyString);
+
+    jstring valueString = (jstring) GetJavaEnv()->CallStaticObjectMethod(storageCls, method, keyString);
+
+    if (mStorageResultDelegate) {
+        mStorageResultDelegate->OnValue(key, GetJavaEnv()->GetStringUTFChars(valueString, 0));
+    }
+}
+
+CHIP_ERROR AndroidDeviceControllerWrapper::GetKeyValue(const char * key, char * value, uint16_t & size)
+{
+    jstring keyString = NULL;
+    CHIP_ERROR err    = CHIP_NO_ERROR;
+    jclass storageCls = GetPersistentStorageClass();
+    jmethodID method  = GetJavaEnv()->GetStaticMethodID(storageCls, "getKeyValue", "(Ljava/lang/String;)Ljava/lang/String;");
+
+    GetJavaEnv()->ExceptionClear();
+
+    err = N2J_NewStringUTF(GetJavaEnv(), key, keyString);
+
+    jstring valueString = (jstring) GetJavaEnv()->CallStaticObjectMethod(storageCls, method, keyString);
+
+    if (value != nullptr) {
+        if (valueString != NULL) {
+            size = strlcpy(value, GetJavaEnv()->GetStringUTFChars(valueString, 0), size);
+        } else {
+            size = strlen(value);
```

Comment:
I don't understand this.  The outparam should be set to the number of bytes placed in `value`.  The incoming bytes in `value` are not initialized.  So this will return a random number in `size`, no?

I'd expect the valueString == NULL case to either return 0 in `size` or error out, depending on what the semantics of that null value are (value is empty string, or there is no value)....

Reply:
My understanding is that `strlcpy` should return the length of the source, which is the second argument.

I've changed the valueString == NULL case to set `size=0`.

Reply:
Yes, I was specifically talking about the `valueString == NULL` case.

Reply:
Oh, I see what's happened here.  The variables got turned around is all...